otp4gb/config.py

```diff
@@ -1,20 +1,110 @@
+"""Functionality for handling the YAML config file."""
+from __future__ import annotations
+
+import datetime
+import json
+import logging
 import os
-import sys
-from yaml import safe_load
+import pathlib
+from typing import Optional
+
+import pydantic
+import caf.toolkit
+
+from otp4gb import cost, parameters, routing, util
+from otp4gb.centroids import Bounds
+
+
+ROOT_DIR = pathlib.Path().absolute()
+BIN_DIR = ROOT_DIR / "bin"
+CONF_DIR = ROOT_DIR / "config"
+ASSET_DIR = ROOT_DIR / "assets"
+LOG_DIR = ROOT_DIR / "logs"
+
+# if you're running on a virtual machine (no virtual memory/page disk)
+# this must not exceed the total amount of RAM.
+PREPARE_MAX_HEAP = os.environ.get("PREPARE_MAX_HEAP", "25G")
+SERVER_MAX_HEAP = os.environ.get("SERVER_MAX_HEAP", "25G")
+LOG = logging.getLogger(__name__)
+
+
+# Pylint incorrectly flags no-member for pydantic.BaseModel
+class TimePeriod(pydantic.BaseModel):  # pylint: disable=no-member
+    """Data required for a single time period."""
+
+    name: str
+    travel_time: datetime.time
+    search_window_minutes: Optional[int] = None
+
+
+class ProcessConfig(caf.toolkit.BaseConfig):
+    """Class for managing (and parsing) the YAML config file."""
+
+    date: datetime.date
+    extents: Bounds
+    osm_file: str
+    gtfs_files: list[str]
+    time_periods: list[TimePeriod]
+    modes: list[list[routing.Mode]]
+    generalised_cost_factors: cost.GeneralisedCostFactors
+    centroids: str
+    destination_centroids: Optional[str] = None
+    iterinary_aggregation_method: cost.AggregationMethod = cost.AggregationMethod.MEAN
+    max_walk_distance: int = 2500
+    number_of_threads: pydantic.conint(ge=0, le=10) = 0
+    no_server: bool = False
+    crowfly_max_distance: Optional[float] = None
+    ruc_lookup: Optional[parameters.RUCLookup] = None
+    irrelevant_destinations: Optional[parameters.IrrelevantDestinations] = None
+    previous_trips: Optional[parameters.PreviousTrips] = None
+
+    # Makes a classmethod not recognised by pylint, hence disabling self check
+    @pydantic.validator("extents", pre=True)
+    def _extents(cls, value):  # pylint: disable=no-self-argument
+        if not isinstance(value, dict):
+            return value
+        return Bounds.from_dict(value)
+
```

Comment:
Updated with [cdca0e3](https://github.com/Transport-for-the-North/otp4gb-py/commit/cdca0e3b0b847d2851d295b0fa19b42be1edf00b)

I see - the more I look at these the more I understand! 

Potentially want to have a chat on Wednesday if possible/time available. Specifically on the general use of them and how to correctly set up.

However, for now lets see if I understand this: 
```
    # Makes a classmethod not recognised by pylint, hence disabling self check
    @pydantic.validator("extents", pre=True)
    def _extents(cls, value):  # pylint: disable=no-self-argument
        if not isinstance(value, dict):
            return value
        return Bounds.from_dict(value)
```

Above is a validator for the `extents` from config.yml. 
As the config specifies the extents as a dict (when loaded from yaml into python object) the validator goes to return Bounds.from_dict(values), the from_dict class methods from the Bounds class. 
Had this just been a list of extents then this list would have been returned immediately.

Additionally, there is no specification on values type (unlike in our example value : str | None) and so, if I understand, it would not throw an error if a set was provided?

Reply:
Yes, that is what that one is doing. That specific validator is used because pydantic can't automatically convert directly to the Bounds class, this validtor has `pre=True` which means it runs before pydantic attempts to do its validation, after this method is ran pydantics automatic validation happens which checks that the returned value is an instance of Bounds and raises an error if not.

More generically these validator methods will be ran by pydantic either before or after its own validtion and if they raise an error pydantic will output those errors in the nice format it does with its own. A lot of validation can be done automatically with normal python type annotations or pydantics more advanced type annotations but if it can't then these custom validators can be used to do whatever custom checks / conversions needed.

Happy to have a discussion about this next week there's quite a lot of useful stuff within pydantic.